Add count of FCDA used in ExtRefs, part of #958

danyill opened the pull request:
I had a go at prototyping counting the number of references to an FCDA in project ExtRefs as suggested in #958. 

It's by no means complete. However, it seems to ruin the plugin's performance by initially making it slow to load after clicking on the editor and then slow to update after doing a binding. It's so laggy I think it would be unacceptable to users.

I would appreciate a little guidance or a hint. How should I best approach this? Can this be made async?

danyill commented:
I guess we could index all ExtRefs and then update the index as they are created or removed. Is that the best approach?

danyill commented:
Hmmmm... incidentally on a large scd file I notice sticky behaviour with the existing implementation (quite apart from the counts).
Possibly performance could be improved by applying this indexing to the "Available to Subscribe" and "Subscribed". 

Using this example file, it takes about 3 seconds between a click on the left-hand side and an update on the right-hand side.

[IOP_2019_HV_v9_ed2.1.zip](https://github.com/openscd/open-scd/files/9484298/IOP_2019_HV_v9_ed2.1.zip)

danyill commented:
I had a look at the performance in Chrome, and it seems like indexing is the way to go. A little wip pushed and now I must learn a little more about events between components. Another day :wink:

danyill commented on src/editors/subscription/later-binding/ext-ref-later-binding-list.ts:
```diff
@@ -266,6 +266,12 @@ export class ExtRefLaterBindingList extends LitElement {
                 this.dispatchEvent(
                   newActionEvent(this.unsubscribe(extRefElement))
                 );
+                this.dispatchEvent(
+                  new CustomEvent('lb-subscription-change', {
+                    bubbles: true,
+                    composed: true,
+                  })
+                );
```
Progress has been made, I think no help :sos: is required

danyill commented:
I am now ready and would be grateful for a review.

This could be made even faster if we updated the count rather than re-indexing on each subscription action, but that seems excessive.

dlabordus commented on src/editors/subscription/later-binding/fcda-later-binding-list.ts:
```diff
@@ -88,6 +98,41 @@ export class FCDALaterBindingList extends LitElement {
+    // index each fcda to display the count promptly
+    const controlElements = this.getControlElements();
+    let fcdas: Element[] = [];
+    controlElements.forEach(controlElement => {
```
I think you can use map/flatMap to convert it and directly assign it the fcdas.

danyill commented on src/editors/subscription/later-binding/fcda-later-binding-list.ts:
```diff
@@ -63,6 +66,13 @@ export class FCDALaterBindingList extends LitElement {
+    const parentDiv = this.closest('.container');
+    if (parentDiv) {
+      parentDiv.addEventListener('lb-subscription-change', () => {
+        this.fcdaCount.clear();
```
> If you clear the Map for every event, what is the benefit of caching the counters?

I clear the map for each subscription change so the read only performance is improved. Otherwise with my naive implementation on a large file there is a lag of several seconds when switching FCDAs.

>I think we can make it smart by sending the identity of the FCDA with the event and only clear that one.

That's a good idea. I thought about doing that but in a less clever way (send a +1 for an FCDA or -1 for an FCDA). 
I worried that eventually the document and state could lose synch and thought the user can accept a lag when doing a subscription (although even that seems too long). For some reason doing what you suggested didn't occur to me 😉 

I would be happy to improve this PR in that direction. That would improve the performance for reading and writing.

> And when building the map not do it in 1 run, but make some getter function that checks if the FCDA is in the map, if not calculate and add, otherwise return the value of the map directly.

Makes sense.

Thank you for this review 😄. I will make improvements soon.

danyill commented on src/editors/subscription/later-binding/fcda-later-binding-list.ts:
```diff
@@ -88,6 +100,53 @@ export class FCDALaterBindingList extends LitElement {
     return [];
   }
 
+  private updateFcdaIndex() {
+    // This method should only be this:
+    //   this.indexFcda(<Element>this.selectedFcdaElement);
+    // but at the key moment this.doc is undefined at the key moment (?).
+    // so I have replaced with fcdaElement.ownerDocument below
+    const fcdaElement = this.selectedFcdaElement!;
+    const iedName = fcdaElement!.closest('IED')!.getAttribute('name') || '';
+    const [ldInst, prefix, lnClass, lnInst, doName, daName] = [
+      'ldInst',
+      'prefix',
+      'lnClass',
+      'lnInst',
+      'doName',
+      'daName',
+    ].map(attr => fcdaElement.getAttribute(attr));
+    // here...
+    const usageCount = Array.from(
+      fcdaElement.ownerDocument.querySelectorAll(
+        `Inputs > ExtRef[iedName="${iedName}"][ldInst="${ldInst}"][prefix="${prefix}"][lnClass="${lnClass}"][lnInst="${lnInst}"][doName="${doName}"][daName="${daName}"]`
+      )
+    ).filter(isPublic).length;
+    this.fcdaIndex.set(identity(fcdaElement), usageCount);
+
+  }
```
I could do with some help as to why I can't access `this.doc` as per the commented-out portion.

danyill commented:
I have made some improvements, added some tests and rebased against main and would be grateful for a further review.

I found that I could not access `this.doc` even though AFAICT it is bound correctly and I can access other class instance properties and would appreciate some guidance there.

dlabordus commented:
@danyill, I like the changes.

We only have one issue, I also did a big refactoring for the Subscriber - Logical Node Plugin (GOOSE/SMV).
So I think we will have several merge conflicts and also need to figure out how to make it work for both.

My proposal is to take your code in a new branch that includes the new plugin and make it work for both plugins (Later and Logical Node Binding). This way I will solve the big conflict I create with refactoring for the new plugin.

Do you find this a good idea?
